heat/core/linalg/solver.py

```diff
@@ -269,3 +270,187 @@ def lanczos(
         V.resplit_(axis=None)
 
     return V, T
+
+
+def solve_triangular(A: DNDarray, b: DNDarray) -> DNDarray:
+    """
+    Solve upper triangular systems of linear equations.
+
+    Input:
+        A - an upper triangular (possibly batched) invertible square (n x n) matrix
+        b - (possibly batched) n x k matrix
+
+    Output:
+        The unique solution x of A * x = b.
+
+    Vectors b have to be given as n x 1 matrices.
+    """
+    if not isinstance(A, DNDarray) or not isinstance(b, DNDarray):
+        raise RuntimeError("Arguments need to be a DNDarrays.")
+    if not A.ndim >= 2:
+        raise RuntimeError("A needs to be a 2D matrix.")
+    if not b.ndim == A.ndim:
+        raise RuntimeError("b needs to be a 2D matrix.")
+    if not A.shape[-2] == A.shape[-1]:
+        raise RuntimeError("A needs to be a square matrix.")
+
+    batch_dim = A.ndim - 2
+    batch_shape = A.shape[:batch_dim]
+
+    if not A.shape[:batch_dim] == b.shape[:batch_dim]:
+        raise RuntimeError("Batch dimensions of A and b must be of the same shape.")
+    if b.split == batch_dim + 1:
+        raise RuntimeError("split=1 is not allowed for the right hand side.")
+    if not b.shape[batch_dim] == A.shape[-1]:
+        raise RuntimeError("Dimension mismath of A and b.")
+
+    if (
+        A.split is not None and A.split < batch_dim or b.split is not None and b.split < batch_dim
+    ):  # batch split
+        if A.split != b.split:
+            raise RuntimeError(
+                "If a split dimension is a batch dimension, A and b must have the same split dimension."
+            )
+    else:
+        if (
+            A.split is not None and b.split is not None
+        ):  # both la dimensions split --> b.split = batch_dim
+            if not all(A.lshape_map[:, A.split] == b.lshape_map[:, batch_dim]):
+                raise RuntimeError("Local arrays of A and b have different sizes.")
```

Comment:
what's the user supposed to do about this?

Reply:
I not sure whether this error can even occur because at this point both dndarrays should have the same size in that axis. (Otherwise lines 294 or 304 should have thrown an exception.)
What do you think @mrfh92 ?

Reply:
Actually, this can happen if at least one of the arrays does not follow the "standard" splitting scheme, e.g., if it is unbalanced for whatever reason.

Reply:
I have updated the error message accordingly.